src/wifi_dpp.rs

```diff
@@ -0,0 +1,451 @@
+        if let Some(dpp) = Self::upgrade_weak_global() {
+            dpp.set_pending_state(state);
+            Ok(())
+        } else {
+            Err(state)
+        }
+    }
+
+    fn wait_for_next_state(&self) -> DppState {
+        self.pending
+            .wait_while_and_get_mut(|state| state.is_none(), |state| state.take().unwrap())
+    }
+
+    fn wait_for_next_state_with_timeout(&self, timeout: Duration) -> Option<DppState> {
+        let (timeout, state_opt) = self.pending.wait_timeout_while_and_get_mut(
+            timeout,
+            |state| state.is_none(),
+            |state| state.take(),
+        );
+        if timeout {
+            None
+        } else {
+            Some(state_opt.unwrap())
+        }
+    }
+
+    fn set_pending_state(&self, state: DppState) {
+        self.pending.get_mut(|s| {
+            *s = Some(state);
+        });
+        self.pending.cvar.notify_all();
+    }
+}
+
+impl Drop for DppInitialized {
+    fn drop(&mut self) {
+        info!("Deinitializing DPP...");
+        unsafe { esp_supp_dpp_deinit() };
+
+        DppInitialized::clear_weak_global();
+    }
+}
+
+pub struct EspWifiDpp<'d, 'w, T> {
+    /// Store the only strong reference to the initialized state in a struct that is guaranteed
+    /// to borrow EspWifi for its lifetime.  This provides the compile-time guarantee that
+    /// we cannot initialize DPP twice.
+    dpp: Arc<DppInitialized>,
+    _phantom: PhantomData<&'d PhantomData<&'w ()>>,
+
+    bootstrapped_data: T,
+}
+
+impl<'d, 'w> EspWifiDpp<'d, 'w, QrCode> {
+    /// Generate a QR code that can be scanned by a mobile phone or other configurator
+    /// to securely provide Wi-Fi credentials.  On success, the caller must invoke
+    /// [::start_listen] to actually start listening.  To wait for the credentials to
+    /// become available, see [DppWait].
+    ///
+    /// Note that [EspWifi] is mutably borrowed for the lifecycle of this object to ensure
+    /// that conflicting usage of the WiFi driver does not occur concurrently with DPP.  It is
+    /// not known the effect this would have and in general it is assumed to be unsafe.
+    ///
+    /// * `wifi` - Mutable borrow for the lifetime of DPP to prevent concurrent usage of the Wi-Fi
+    /// driver.
+    /// * `channels` - List of channels to listen for DPP auth messages.
+    /// * `key` - (Optional) NIST P-256 private key to use when generating the QR code.  This can
+    /// be useful for example so that the QR code can be printed and distributed with the device.
+    /// If omitted, a unique private key is generated on each invocation.  Do not include PEM
+    /// or DER formatting data as it will be added automatically depending on which version of
+    /// ESP-IDF is being used.
+    /// * `associated_data` - (Optional) Arbitrary extra information to include with the QR
+    /// code that may be relevant to the configurator.
+    pub fn generate_qrcode(
```

Comment:
Why not naming it simply... `new`?

Reply:
In the future I expect there to be other ways to bootstrap DPP, as DPP itself supports QR code, proof of knowledge, NFC URI, and even Bluetooth in later revisions.  Espressif's API even supports this idea through an awkward enum, though I'd prefer the Rust API expose this with new fn's like `fn using_pkex(...)`, `fn using_ble(...)`, etc.